_posts/2016-12-20-sql-functions-vs-active-record.markdown

```diff
@@ -0,0 +1,67 @@
+
+A six-hour bottleneck at its slowest, to be exact. With several key stakeholders spread out across the Northeast requiring at least one report every day, the slowness of the process necessitated late-night, early morning or weekend report generation to avoid potential side effects slowing down regular business operations.
+
+Though the code itself, by nature of it being Ruby, was semantically easy to read and understand, ActiveRecord, no matter how much we refactored and eager-loaded, was proving to not be a sustainable solution as we continued to grow our portfolio and our associated records increased exponentially.
+
+### Enter SQL ###
+
+At first, my team and I discussed a variety of solutions, such as porting out the reporting to another application in a more math-friendly language such as Elixir, to incorporating an already-in-use Solr solution. The overhead involved in getting everything up and running proved to be too large, and without any guarantee of a meaningful performance increase.
+
+In the end, we settled on a lower-level language solution — Custom SQL functions. By translating all of the calculations that were initially in the Ruby module to custom SQL functions, we were able to avoid iterating over hundreds of thousands records and perform a single query off the database. 
+
+I first started by building out a query that would establish a base unit on which all of our other calculations would be based across the rest of the report. In this situation, it was the sum total of all of our payments for a scoped range set by the user:
+
+```
+CREATE OR REPLACE FUNCTION total_payments_for_range(
+    payment_start_date date, 
+    payment_end_date date
+)
+    RETURNS TABLE (
+      total_payments_for_range numeric,
+      advance_id integer
+    ) AS $$
+    SELECT
+      COALESCE(sum(p.actual_holdback_amount), 0) as total_payments_for_range,
+      id as advance_id
+    FROM advances a
+    LEFT JOIN (
+      SELECT
+        actual_holdback_amount,
+        advance_id
+      FROM payments p
+      WHERE p.payment_date >= payment_start_date AND p.payment_date <= payment_end_date
+    ) as p
+    ON p.advance_id = a.id
+    GROUP BY a.id
+```
+
+By structuring all of our subsequent calculations around this base query, we were able to bring down our total report time to around 27 seconds on the first iteration of the SQL report function. Further refactors to remove my myriad structural problems, mostly duplicated or redundant queries, I was finally able to get our current build down to... *drumroll please*
```

Comment:
It's myriad. Time for some fun grammar trivia...

Myriad in Greek is 10,000. 

Thus:
>10,000 men == Myriad men
>10,000 of men == Myriad of men

Technically, both usages of `myriad men` and `myriad of men` are correct, but the more literally elegant usage is always as an adjective. Most copy editors will change it to the adjective usage.